content/300-guides/300-upgrade-guides/200-upgrading-versions/030-upgrading-to-prisma-4.mdx

```diff
@@ -0,0 +1,431 @@
+
+See our [system requirements](/reference/system-requirements) for all minimum version requirements.
+
+### Raw query type mapping: scalar values are now deserialized as their correct JavaScript types
+
+In versions 3.14.x and 3.15.x, [raw query type mapping](/concepts/components/prisma-client/raw-database-access#raw-query-type-mapping) was available with the Preview feature `improvedQueryRaw`. In version 4.0.0, we have made raw query type mapping Generally Available. You do not need to use `improvedQueryRaw` to get this functionality in versions 4.0.0 and later.
+
+Raw queries now deserialize scalar values to their corresponding JavaScript types. Note that Prisma infers types from the values themselves and not from the Prisma Schema types.
+
+Example query and response:
+
+```ts
+const res =
+  await prisma.$queryRaw`SELECT bigint, bytes, decimal, date FROM "Table";`
+console.log(res) // [{ bigint: BigInt("123"), bytes: Buffer.from([1, 2]), decimal: new Prisma.Decimal("12.34"), date: Date("<some_date>") }]
+```
+
+#### Upgrade path
+
+From version 4.0.0, some data types returned by `queryRaw` or `queryRawUnsafe` are different, as follows:
+
+| Data type  | Before version 4.0.0  | From version 4.0.0    |
+| ---------- | --------------------- | --------------------- |
+| `DateTime` | Returned as `String`  | Returned as `Date`    |
+| `Numeric`  | Returned as `Float`   | Returned as `Decimal` |
+| `Bytes`    | Returned as `String`  | Returned as `Buffer`  |
+| `Int64`    | Returned as `Integer` | Returned as `BigInt`  |
+
+If you use `queryRaw` or `queryRawUnsafe` to return any of the above data types, then you must change your code to handle the new types.
+
+For example, if you return `DateTime` data, then you need to take into account the following:
+
+- You no longer need to manually instantiate a `DateTime` object for the returned data.
+- If your code currently uses the returned `String` data, then you now need to convert the `DateTime` object to a `String`.
+
+You must make equivalent code changes for the other data types in the table above.
+
+### Raw query mapping: PostgreSQL type-casts
+
+In versions 3.14.x and 3.15.x, [raw query type mapping](/concepts/components/prisma-client/raw-database-access#raw-query-type-mapping) was available with the Preview feature `improvedQueryRaw`. In version 4.0.0, we have made raw query type mapping Generally Available. You do not need to use `improvedQueryRaw` to get this functionality in versions 4.0.0 and later.
+
+Before version 4.0.0, many PostgreSQL type-casts did not work. We have tightened the type coercion rules so that all type-casts now work. As a result, some implicit casts now fail.
+
+#### Upgrade path
+
+We recommend that you re-test your use of `$queryRaw` to ensure that the types you pass into your raw queries match the types that PostgreSQL expects.
+
+For example, in version 4.0.0, the following query fails:
+
+```js
+await prisma.$queryRaw`select length(${42});
```

Comment:
This code example is missing a closing backtick, which breaks the syntax highlighting on GitHub and also probably the rendering.

Reply:
Probably:
```suggestion
await prisma.$queryRaw`select length(${42});`
```